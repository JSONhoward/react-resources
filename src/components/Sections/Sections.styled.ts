import styled from 'styled-components';

export const StyledSections = styled('div')`
display: grid;
grid-template-columns: repeat(auto-fit, minmax(280px, 1fr));
gap: .5rem;
justify-items: center;
padding: .5rem;
`

export  const SectionTitle = styled('p')`
text-align: center;
color: #6F6F6F;
font-size: 1.5rem;
`

export const Img = styled('img')<{state: string, circle: boolean}>`
position: absolute;
top: 50%;
left: 50%;
transform: translate(-50%,-50%);
border-radius: ${({circle}) => circle ? '50%' : 'none'};
height: 10rem;
width: 10rem;
box-shadow: 2px 2px 5px rgba(0,0,0,.5);
transition: all 1s ease-in-out;
opacity: ${({state}) => {
    const obj: {[key: string]: number} = {
        'entering': 1,
        'entered': 1,
        'exiting': 0,
        'exited': 0
    }
    return obj[state]
}};

&:hover {
    height: 12rem;
    width: 12rem;
}
`

export const SectionDescription1 = styled('div')`
position: relative;
display: flex;
align-items: center;
height: 50%;
width: 100%;
`

export const SectionDescription2 = styled('div')`
position: relative;
display: flex;
align-items: center;
height: 50%;
width: 100%;
`

export const SectionDescription3 = styled('div')`
position: relative;
display: flex;
align-items: center;
height: 50%;
width: 100%;
`

export const Section = styled('div')`
position: relative;
display: flex;
flex-direction: column;
justify-content: space-evenly;
height: 300px;
width: 100%;
`