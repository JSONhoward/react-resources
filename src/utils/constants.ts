import cleverP from '../images/clever-programmer-icon.jpg'
import devEd from '../images/dev-ed-icon.jpg'
import fireship from '../images/fireship-icon.jpg'
import freeCodeCamp from '../images/free-code-camp-icon.jpg'
import netNinja from '../images/net-ninja-icon.jpg'
import webDevSimplified from '../images/web-dev-simplified-icon.jpg'
import rwieruch from '../images/rwieruch-icon.svg'
import sitePoint from '../images/sitepoint.svg'
import smashingMagazine from '../images/smashing-magazine.svg'
import codeCademy from '../images/codecademy.svg'
import udemy from '../images/udemy-icon.svg'
import wesBos from '../images/wes-bos-icon.png'
import medium from '../images/medium-icon.png'
import devTo from '../images/dev-to.svg'

export const webDevChannels: { name: string, id: string, icon: string }[] = [
    {
        id: 'UCFbNIlppjAuEX4znoulh0Cw',
        name: 'Web Dev Simplified',
        icon: webDevSimplified
    },
    {
        id: 'UClb90NQQcskPUGDIXsQEz5Q',
        name: 'Dev Ed',
        icon: devEd
    },
    {
        id: 'UCW5YeuERMmlnqo4oq8vwUpg',
        name: 'Net Ninja',
        icon: netNinja
    },
    {
        id: 'UCsBjURrPoezykLs9EqgamOA',
        name: 'Fireship',
        icon: fireship
    },
    {
        id: 'UC8butISFwT-Wl7EV0hUK0BQ',
        name: 'Free Code Camp',
        icon: freeCodeCamp
    },
    {
        id: 'UCqrILQNl5Ed9Dz6CGMyvMTQ',
        name: 'Clever Programmer',
        icon: cleverP
    }
]

export const mockArray: {id: string, url: string, title: string, description: string, thumb: string}[] = [
    {
        id: 'someId',
        url: 'https://www.youtube.com/watch?v=Faz6jkFTaa0',
        title: 'title goes here',
        description: 'lorem ipsum',
        thumb: `https://via.placeholder.com/300x200`
    },
    {
        id: 'someId',
        url: 'https://www.youtube.com/watch?v=Faz6jkFTaa0',
        title: 'title goes here',
        description: 'lorem ipsum',
        thumb: `https://via.placeholder.com/300x200`
    }, {
        id: 'someId',
        url: 'https://www.youtube.com/watch?v=Faz6jkFTaa0',
        title: 'title goes here',
        description: 'lorem ipsum',
        thumb: `https://via.placeholder.com/300x200`
    }, {
        id: 'someId',
        url: 'https://www.youtube.com/watch?v=Faz6jkFTaa0',
        title: 'title goes here',
        description: 'lorem ipsum',
        thumb: `https://via.placeholder.com/300x200`
    },
    {
        id: 'someId',
        url: 'https://www.youtube.com/watch?v=Faz6jkFTaa0',
        title: 'title goes here',
        description: 'lorem ipsum',
        thumb: `https://via.placeholder.com/300x200`
    }
]

export const onlineCourses: {platform: string, name: string, url: string, icon: string}[] = [
    {
        platform: 'FreeCodeCamp',
        name: 'Learn React in 1 Hour by Building a Movie Search App',
        url: 'https://www.freecodecamp.org/news/learn-react-in-1-hour-by-building-a-movie-search-app/',
        icon: freeCodeCamp
    },
    {
        platform: 'Udemy',
        name: 'React - The Complete Guide (incl Hooks, React Router, Redux)',
        url: 'https://www.udemy.com/course/react-the-complete-guide-incl-redux/',
        icon: udemy
    },
    {
        platform: 'CodeCademy',
        name: 'Learn React.js: Part I/Part II',
        url: 'https://www.codecademy.com/learn/react-101',
        icon: codeCademy
    },
    {
        platform: 'none',
        name: 'Fullstack Advanced React & GraphQL',
        url: 'https://advancedreact.com/',
        icon: wesBos
    }
]

export const onlineBlogs: {name: string, url: string, icon: string}[] = [
    {
        name: 'Medium',
        url: 'https://medium.com/',
        icon: medium
    },
    {
        name: 'Dev.to',
        url: 'https://dev.to/',
        icon: devTo
    },
    {
        name: 'Sitepoint',
        url: 'https://www.sitepoint.com/',
        icon: sitePoint
    },
    {
        name: 'Smashing Magazine',
        url: 'https://www.smashingmagazine.com/',
        icon: smashingMagazine
    },
    {
        name: 'RobinWieruch',
        url: 'https://www.robinwieruch.de/blog',
        icon: rwieruch
    }
]