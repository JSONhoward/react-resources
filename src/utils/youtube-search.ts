const YOUTUBE_SEARCH_URL = (channel: string, limit: string): string => `https://www.googleapis.com/youtube/v3/search?part=snippet&channelId=${channel}${limit}&q=react&type=video&key=${process.env.REACT_APP_YOUTUBE_API}`

export const getVideoList = async (channel: string, limit = 50): Promise<Video[]> => {
    const response = await fetch(YOUTUBE_SEARCH_URL(channel, `&maxResults=${limit.toString()}`))
    const data = await response.json()

    const videos: Video[] = await data['items'].map((result: { id: { videoId: any }; snippet: { title: any; description: any; thumbnails: { high: { url: any } } } }) => {
        return {
            id: result.id.videoId,
            url: `http://www.youtube.com/watch?v=${result.id.videoId}`,
            title: result.snippet.title,
            description: result.snippet.description,
            thumb: result.snippet.thumbnails.high.url
        }
    })

    return videos
}