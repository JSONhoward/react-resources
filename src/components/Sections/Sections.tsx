import React from 'react'
import { StyledSections, SectionDescription, SectionTitle, Section, Img } from './Sections.styled';
import { webDevChannels, onlineCourses, onlineBlogs } from '../../utils/constants';
import { Transition } from 'react-transition-group';
import { Link } from 'react-router-dom';

const Sections = () => {
    const [tick, setTick] = React.useState(0)

    React.useEffect(() => {
        const ticker = setInterval(() => setTick(prev => prev + 1), 2000)

        return () => clearInterval(ticker)
    }, [])

    const channelImages = webDevChannels.map((_, i) => {
        return (
            <Transition key={i} in={i === tick % webDevChannels.length} timeout={1500}>
                {
                    state => (
                        <Link data-testid={`sectionChannelLink${i}`} to='/channels'>
                            <Img data-testid={`sectionChannelIcon${i}`} circle={true} state={state} src={webDevChannels[i].icon} alt={webDevChannels[i].name} />
                        </Link>
                    )
                }
            </Transition>
        )
    })

    const courseImages = onlineCourses.map((_, i) => {
        return (
            <Transition key={i} in={i === tick % onlineCourses.length} timeout={1500}>
                {
                    state => (
                        <Link data-testid={`sectionCourseLink${i}`} to='/courses'>
                            <Img data-testid={`sectionCourseIcon${i}`} circle={false} state={state} src={onlineCourses[i].icon} alt={onlineCourses[i].name} />
                        </Link>
                    )
                }
            </Transition>
        )
    })

    const blogImages = onlineBlogs.map((_, i) => {
        return (
            <Transition key={i} in={i === tick % onlineBlogs.length} timeout={1500}>
                {
                    state => (
                        <Link data-testid={`sectionBlogLink${i}`} to='/blogs'>
                            <Img data-testid={`sectionBlogIcon${i}`} circle={true} state={state} src={onlineBlogs[i].icon} alt={onlineBlogs[i].name} />
                        </Link>
                    )
                }
            </Transition>
        )
    })

    return (
        <StyledSections>
            <Section>
                <SectionTitle>
                    <Link to='/channels'>
                        Youtube Channels
                    </Link>
                </SectionTitle>
                <SectionDescription>
                    {channelImages}
                </SectionDescription>
            </Section>
            <Section>
                <SectionTitle>
                    <Link to='/courses'>
                        Online Courses
                    </Link>
                </SectionTitle>
                <SectionDescription>
                    {courseImages}
                </SectionDescription>
            </Section>
            <Section>
                <SectionTitle>
                    <Link to='/blogs'>
                        Blogs
                    </Link>
                </SectionTitle>
                <SectionDescription>
                    {blogImages}
                </SectionDescription>
            </Section>

        </StyledSections>
    )
}

export default Sections
