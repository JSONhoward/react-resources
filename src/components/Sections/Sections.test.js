import React from 'react'
import {BrowserRouter as Router} from 'react-router-dom'
import {create} from 'react-test-renderer'
import {render, screen} from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import Sections from './'
import { onlineBlogs, onlineCourses, webDevChannels } from '../../utils/constants'

describe('Sections Tests', () => {
    test('Snapshot Test', () => {
        const tree = create(<Router><Sections /></Router>).toJSON()

        expect(tree).toMatchSnapshot()
    })

    test('Renders each channel icon', () => {
        render(<Router><Sections /></Router>)

        webDevChannels.forEach(({icon, name}, i) => {
            expect(screen.getByTestId(`sectionChannelLink${i}`)).toHaveAttribute('href', '/channels')
            expect(screen.getByTestId(`sectionChannelIcon${i}`)).toHaveAttribute('src', icon)
            expect(screen.getByTestId(`sectionChannelIcon${i}`)).toHaveAttribute('alt', name)
        })
    })

    test('Renders each course icon', () => {
        render(<Router><Sections /></Router>)

        onlineCourses.forEach(({icon, name}, i) => {
            expect(screen.getByTestId(`sectionCourseLink${i}`)).toHaveAttribute('href', '/courses')
            expect(screen.getByTestId(`sectionCourseIcon${i}`)).toHaveAttribute('src', icon)
            expect(screen.getByTestId(`sectionCourseIcon${i}`)).toHaveAttribute('alt', name)
        })
    })

    test('Renders each blog icon', () => {
        render(<Router><Sections /></Router>)

        onlineBlogs.forEach(({icon, name}, i) => {
            expect(screen.getByTestId(`sectionBlogLink${i}`)).toHaveAttribute('href', '/blogs')
            expect(screen.getByTestId(`sectionBlogIcon${i}`)).toHaveAttribute('src', icon)
            expect(screen.getByTestId(`sectionBlogIcon${i}`)).toHaveAttribute('alt', name)
        })
    })
})