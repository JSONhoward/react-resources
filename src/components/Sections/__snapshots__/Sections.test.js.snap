// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sections Tests Snapshot Test 1`] = `
<div
  className="sc-bdnylx fuQbuA"
>
  <div
    className="sc-eCApGN ixvGpU"
  >
    <p
      className="sc-gtssRu dzmwPh"
    >
      <a
        href="/channels"
        onClick={[Function]}
      >
        Youtube Channels
      </a>
    </p>
    <div
      className="sc-hKFyIo cUiZjG"
    >
      <a
        data-testid="sectionChannelLink0"
        href="/channels"
        onClick={[Function]}
      >
        <img
          alt="Web Dev Simplified"
          className="sc-dlnjPT jxHESt"
          data-testid="sectionChannelIcon0"
          src="web-dev-simplified-icon.jpg"
        />
      </a>
      <a
        data-testid="sectionChannelLink1"
        href="/channels"
        onClick={[Function]}
      >
        <img
          alt="Dev Ed"
          className="sc-dlnjPT eMTUHU"
          data-testid="sectionChannelIcon1"
          src="dev-ed-icon.jpg"
        />
      </a>
      <a
        data-testid="sectionChannelLink2"
        href="/channels"
        onClick={[Function]}
      >
        <img
          alt="Net Ninja"
          className="sc-dlnjPT eMTUHU"
          data-testid="sectionChannelIcon2"
          src="net-ninja-icon.jpg"
        />
      </a>
      <a
        data-testid="sectionChannelLink3"
        href="/channels"
        onClick={[Function]}
      >
        <img
          alt="Fireship"
          className="sc-dlnjPT eMTUHU"
          data-testid="sectionChannelIcon3"
          src="fireship-icon.jpg"
        />
      </a>
      <a
        data-testid="sectionChannelLink4"
        href="/channels"
        onClick={[Function]}
      >
        <img
          alt="Free Code Camp"
          className="sc-dlnjPT eMTUHU"
          data-testid="sectionChannelIcon4"
          src="free-code-camp-icon.jpg"
        />
      </a>
      <a
        data-testid="sectionChannelLink5"
        href="/channels"
        onClick={[Function]}
      >
        <img
          alt="Clever Programmer"
          className="sc-dlnjPT eMTUHU"
          data-testid="sectionChannelIcon5"
          src="clever-programmer-icon.jpg"
        />
      </a>
    </div>
  </div>
  <div
    className="sc-eCApGN ixvGpU"
  >
    <p
      className="sc-gtssRu dzmwPh"
    >
      <a
        href="/courses"
        onClick={[Function]}
      >
        Online Courses
      </a>
    </p>
    <div
      className="sc-hKFyIo cUiZjG"
    >
      <a
        data-testid="sectionCourseLink0"
        href="/courses"
        onClick={[Function]}
      >
        <img
          alt="Learn React in 1 Hour by Building a Movie Search App"
          className="sc-dlnjPT gNWSff"
          data-testid="sectionCourseIcon0"
          src="free-code-camp-icon.jpg"
        />
      </a>
      <a
        data-testid="sectionCourseLink1"
        href="/courses"
        onClick={[Function]}
      >
        <img
          alt="React - The Complete Guide (incl Hooks, React Router, Redux)"
          className="sc-dlnjPT dUntHe"
          data-testid="sectionCourseIcon1"
          src="udemy-icon.svg"
        />
      </a>
      <a
        data-testid="sectionCourseLink2"
        href="/courses"
        onClick={[Function]}
      >
        <img
          alt="Learn React.js: Part I/Part II"
          className="sc-dlnjPT dUntHe"
          data-testid="sectionCourseIcon2"
          src="codecademy.svg"
        />
      </a>
      <a
        data-testid="sectionCourseLink3"
        href="/courses"
        onClick={[Function]}
      >
        <img
          alt="Fullstack Advanced React & GraphQL"
          className="sc-dlnjPT dUntHe"
          data-testid="sectionCourseIcon3"
          src="wes-bos-icon.png"
        />
      </a>
    </div>
  </div>
  <div
    className="sc-eCApGN ixvGpU"
  >
    <p
      className="sc-gtssRu dzmwPh"
    >
      <a
        href="/blogs"
        onClick={[Function]}
      >
        Blogs
      </a>
    </p>
    <div
      className="sc-hKFyIo cUiZjG"
    >
      <a
        data-testid="sectionBlogLink0"
        href="/blogs"
        onClick={[Function]}
      >
        <img
          alt="Medium"
          className="sc-dlnjPT jxHESt"
          data-testid="sectionBlogIcon0"
          src="medium-icon.png"
        />
      </a>
      <a
        data-testid="sectionBlogLink1"
        href="/blogs"
        onClick={[Function]}
      >
        <img
          alt="Dev.to"
          className="sc-dlnjPT eMTUHU"
          data-testid="sectionBlogIcon1"
          src="dev-to.svg"
        />
      </a>
      <a
        data-testid="sectionBlogLink2"
        href="/blogs"
        onClick={[Function]}
      >
        <img
          alt="Sitepoint"
          className="sc-dlnjPT eMTUHU"
          data-testid="sectionBlogIcon2"
          src="sitepoint.svg"
        />
      </a>
      <a
        data-testid="sectionBlogLink3"
        href="/blogs"
        onClick={[Function]}
      >
        <img
          alt="Smashing Magazine"
          className="sc-dlnjPT eMTUHU"
          data-testid="sectionBlogIcon3"
          src="smashing-magazine.svg"
        />
      </a>
      <a
        data-testid="sectionBlogLink4"
        href="/blogs"
        onClick={[Function]}
      >
        <img
          alt="RobinWieruch"
          className="sc-dlnjPT eMTUHU"
          data-testid="sectionBlogIcon4"
          src="rwieruch-icon.svg"
        />
      </a>
    </div>
  </div>
</div>
`;
